src/rt/sched/cown.h

```diff
@@ -771,11 +771,99 @@ namespace verona::rt
         Scheduler::record_inflight_message();
       }
 
+      if ((sched != nullptr) && (sched->message_body != nullptr))
+        backpressure_scan(*sched->message_body, *body);
+
       // Try to acquire as many cowns as possible without rescheduling,
       // starting from the beginning.
       fast_send(body, epoch);
     }
 
+    /**
+     * Return true if any senders are overloaded or if any sender is in the
+     * receiver set.
+     */
+    static inline bool backpressure_priority(
```

Comment:
I find this method name opaque and confusing.

Reply:
Yes, but still sounds confusing.

Reply:
I think the whole logic here is confusing. There is no relation between which iteration returned `true` here and which receiver will be unmuted later.

From earlier conversations, it seems there can only be two cases: one sender, multiple receivers or multiple senders and one receiver, but the code seems to multiple of both.

Perhaps factoring some of the code out to smaller functions with more accurate names and combining them in the logical way could clear things up.

Reply:
There is no relationship between where this function returns true and what receiver would be unmuted. Requesting that a receiver be unmuted before determining if it should be seems incorrect to me, and I would prefer an additional loop over the short array of receivers than an allocation to track which ones might possibly require unmuting.

The name of this function should line up with the description of "priority" from the summary comment in backpressure.h. If the description or the name "priority" are unclear, then that needs to be addressed.

There can be any amount of senders and receivers for a multi-message.

Reply:
> Requesting that a receiver be unmuted before determining if it should be seems incorrect to me

That's not what I meant.

If there could be multiple senders and multiple receivers, and if any sender could return true but only receivers from the same sender, then unmuting *any* sender that is overloaded seems to not match my interpretation of safety.

Then again, it might be that I'm misinterpreting what the semantics are, or what the code does, which means the code is slightly confusing, at least to me, personally.

My proposal was only to clarify what the code is doing. I find it helpful if I break the functionality in small pieces when the code is not one clear step after the other, even if it's only called from one site. Either that, or some expressive comment before each block, but that's less preferable than a nice function name. :)

Reply:
Sorry, I think I'm confused about your confusion :)

> if any sender could return true but only receivers from the same sender

What do you mean by "but only receivers from the same sender"?

> then unmuting any sender that is overloaded seems to not match my interpretation of safety

Senders would never be unmuted, because they cannot be muted. A muted cown cannot be running an action and therefore cannot be sending messages.

Reply:
> Sorry, I think I'm confused about your confusion :)

:D

> What do you mean by "but only receivers from the same sender"?

I meant:
```
    if (receiver == sender)
        return true;
```

> > then unmuting any sender that is overloaded seems to not match my interpretation of safety
> 
> Senders would never be unmuted, because they cannot be muted. A muted cown cannot be running an action and therefore cannot be sending messages.

Sorry again, "unmuting any receivers", which is what the code is doing when this function returns true.